plugin/plugins/filesystem/nodes/base.go

```diff
@@ -0,0 +1,148 @@
+package fsnodes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/djdv/p9/p9"
+	"github.com/djdv/p9/unimplfs"
+	files "github.com/ipfs/go-ipfs-files"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+const ( //device - attempts to comply with standard multicodec table
+	dMemory = 0x2f // generic "path"
+	dIPFS   = 0xe4
+)
+
+var _ p9.File = (*Base)(nil)
+
+// Base is a foundational file system node that provides common file metadata as well as stubs for unimplemented methods
+type Base struct {
+	// Provide stubs for unimplemented methods
+	unimplfs.NoopFile
+	p9.DefaultWalkGetAttr
+
+	// Storage for file's metadata
+	Qid      p9.QID
+	meta     p9.Attr
+	metaMask p9.AttrMask
+
+	// The parent context should be set prior to calling attach (usually in a consturctor)
+	// The Base filesystemCtx is derived from the parent context during Base.Attach
+	// and should be valid for the lifetime of the file system
+	// The fs context should be used to derive operation specific contexts from during calls
+	// A cancel should cascade down, and invalidate all derived contexts
+	// A call to fs.Close should leave no operations lingering
+	parentCtx        context.Context
+	filesystemCtx    context.Context
+	filesystemCancel context.CancelFunc
+	Logger           logging.EventLogger
+
+	parent p9.File // parent must be set, it is used to handle ".." requests; nodes without a parent must point back to themselves
+	child  p9.File // child is an optional field, used to hand off walk requests to another file system
+	root   bool    // should be set to true on Attach if parent == self, this triggers the filesystemCancel on Close
+	open   bool    // should be set to true on Open and checked during Walk; do not walk open references walk(5)
+}
+
+// IPFSBase is much like Base but extends it to hold IPFS specific metadata
+type IPFSBase struct {
+	Base
+
+	Path corepath.Resolved
+	core coreiface.CoreAPI
+
+	// you will typically want to derive a context from the Base context within one operation (like Open)
+	// use it with the CoreAPI for something (like Get)
+	// and cancel it in another operation (like Close)
+	// that pointer should be stored here between calls
+	operationsCancel context.CancelFunc
+
+	// operation handle storage
+	file      files.File
+	directory *directoryStream
+}
+
+// Base Attach should be called by all supersets during their Attach
+// to initialize the file system context
+func (b *Base) Attach() (p9.File, error) {
+	if b.parentCtx == nil {
+		return nil, errors.New("Parent context was not set, no way to derive")
```

Comment:
We should probably _panic_ here. This is a logic bug.

Reply:
Things have changed where panics will be triggered by the runtime for operations like this.
The constructors themselves will panic if requirements are not met (passing in a `nil` context or MFS root for example), and if the object somehow becomes malformed it is likely to panic during operation for similar reasons (unable to derive during `Walk`, unable to talk to the core, unable to access metadata, etc.).

Excluding the constructors, the library above us handles panic recovery gracefully. But this is never expected to happen.